org.alloytools.alloy.classic.provider/src/main/java/org/alloytools/alloy/classic/provider/Relation.java

```diff
@@ -0,0 +1,202 @@
+package org.alloytools.alloy.classic.provider;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import org.alloytools.alloy.core.api.IAtom;
+import org.alloytools.alloy.core.api.IRelation;
+import org.alloytools.alloy.core.api.ITuple;
+import org.alloytools.alloy.core.api.Solution;
+
+public class Relation implements IRelation {
+
+    final Solution solution;
+    final int      arity;
+    final ITuple[] tuples;
+
+    Relation(Solution solution, int arity, Tuple[] tuples) {
+        this.solution = solution;
+        this.tuples = tuples;
+        this.arity = arity;
+    }
+
+    public Relation(Solution solution, int arity, List< ? extends IAtom> atoms) {
+        this(solution, arity, toTuples(solution, arity, atoms));
+    }
+
+    @Override
+    public int arity() {
+        return arity;
+    }
+
+    @Override
+    public IRelation join(IRelation right) {
```

Comment:
are there any unit tests for this implementation? (as well as for other methods in this class)